// eas-cli/build/metadata/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAppleWriter = exports.createAppleReader = exports.validateConfig = exports.loadConfigAsync = exports.getStaticConfigFilePath = void 0;
const tslib_1 = require("tslib");
const ajv_1 = tslib_1.__importDefault(require("ajv"));
const assert_1 = tslib_1.__importDefault(require("assert"));
const fs_extra_1 = tslib_1.__importDefault(require("fs-extra"));
const path_1 = tslib_1.__importDefault(require("path"));
const reader_1 = require("./apple/config/reader");
const writer_1 = require("./apple/config/writer");
const errors_1 = require("./errors");
/**
 * Resolve the dynamic config from the user.
 * It supports methods, async methods, or objects (json).
 */
async function resolveDynamicConfigAsync(configFile) {
    const userConfigOrFunction = await Promise.resolve().then(() => tslib_1.__importStar(require(configFile))).then(file => { var _a; return (_a = file.default) !== null && _a !== void 0 ? _a : file; });
    return typeof userConfigOrFunction === 'function'
        ? await userConfigOrFunction()
        : userConfigOrFunction;
}
/**
 * Get the static configuration file path, based on the metadata context.
 * This uses any custom name provided, but swaps out the extension for `.json`.
 */
function getStaticConfigFilePath({ projectDir, metadataPath, }) {
    const configFile = path_1.default.join(projectDir, metadataPath);
    const configExtension = path_1.default.extname(configFile);
    return path_1.default.join(projectDir, `${path_1.default.basename(configFile, configExtension)}.json`);
}
exports.getStaticConfigFilePath = getStaticConfigFilePath;
/**
 * Load the store configuration from a metadata context.
 * This can load `.json` and `.js` config files, using `require`.
 * It throws MetadataValidationErrors when the file doesn't exist, or contains errors.
 * The user is prompted to try anyway when errors are found.
 */
async function loadConfigAsync({ projectDir, metadataPath, skipValidation = false, }) {
    const configFile = path_1.default.join(projectDir, metadataPath);
    if (!(await fs_extra_1.default.pathExists(configFile))) {
        throw new errors_1.MetadataValidationError(`Metadata store config file not found: "${configFile}"`);
    }
    const configData = await resolveDynamicConfigAsync(configFile);
    if (!skipValidation) {
        const { valid, errors: validationErrors } = validateConfig(configData);
        if (!valid) {
            throw new errors_1.MetadataValidationError(`Metadata store config errors found`, validationErrors);
        }
    }
    return configData;
}
exports.loadConfigAsync = loadConfigAsync;
/**
 * Run the JSON Schema validation to normalize defaults and flag early config errors.
 * This includes validating the known store limitations for every configurable property.
 */
function validateConfig(config) {
    const validator = new ajv_1.default({ allErrors: true, useDefaults: true })
        .addMetaSchema(require('ajv/lib/refs/json-schema-draft-06.json'))
        .compile(require('../../schema/metadata-0.json'));
    const valid = validator(config);
    return { valid, errors: validator.errors || [] };
}
exports.validateConfig = validateConfig;
/** Create a versioned deserializer to fetch App Store data from the store configuration. */
function createAppleReader(config) {
    (0, assert_1.default)(config.configVersion === 0, 'Unsupported store configuration version');
    (0, assert_1.default)(config.apple, 'No apple configuration found');
    return new reader_1.AppleConfigReader(config.apple);
}
exports.createAppleReader = createAppleReader;
/** Create the serializer to write the App Store to the store configuration. */
function createAppleWriter() {
    return new writer_1.AppleConfigWriter();
}
exports.createAppleWriter = createAppleWriter;
