// eas-cli/build/build/ios/credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ensureIosCredentialsAsync = void 0;
const tslib_1 = require("tslib");
const eas_build_job_1 = require("@expo/eas-build-job");
const IosCredentialsProvider_1 = tslib_1.__importDefault(require("../../credentials/ios/IosCredentialsProvider"));
const BuildCredentialsUtils_1 = require("../../credentials/ios/actions/BuildCredentialsUtils");
const credentials_1 = require("../utils/credentials");
async function ensureIosCredentialsAsync(buildCtx, targets) {
    var _a;
    if (!shouldProvideCredentials(buildCtx)) {
        return;
    }
    const provider = new IosCredentialsProvider_1.default(buildCtx.credentialsCtx, {
        app: (0, BuildCredentialsUtils_1.getAppFromContext)(buildCtx.credentialsCtx),
        targets,
        distribution: (_a = buildCtx.buildProfile.distribution) !== null && _a !== void 0 ? _a : 'store',
        enterpriseProvisioning: buildCtx.buildProfile.enterpriseProvisioning,
    });
    const { credentialsSource } = buildCtx.buildProfile;
    (0, credentials_1.logCredentialsSource)(credentialsSource, eas_build_job_1.Platform.IOS);
    return {
        credentials: await provider.getCredentialsAsync(credentialsSource),
        source: credentialsSource,
    };
}
exports.ensureIosCredentialsAsync = ensureIosCredentialsAsync;
function shouldProvideCredentials(buildCtx) {
    return !buildCtx.buildProfile.simulator;
}
