// eas-cli/build/graphql/queries/UpdateQuery.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateQuery = void 0;
const tslib_1 = require("tslib");
const graphql_tag_1 = tslib_1.__importDefault(require("graphql-tag"));
const client_1 = require("../client");
exports.UpdateQuery = {
    async viewAllAsync({ appId, limit, offset }) {
        return (0, client_1.withErrorHandlingAsync)(client_1.graphqlClient
            .query((0, graphql_tag_1.default) `
            query ViewAllUpdates($appId: String!, $limit: Int!, $offset: Int!) {
              app {
                byId(appId: $appId) {
                  id
                  updates(limit: $limit, offset: $offset) {
                    id
                    group
                    message
                    createdAt
                    runtimeVersion
                    platform
                    actor {
                      id
                      ... on User {
                        username
                      }
                      ... on Robot {
                        firstName
                      }
                    }
                    branch {
                      id
                      name
                    }
                  }
                }
              }
            }
          `, {
            appId,
            limit,
            offset,
        }, { additionalTypenames: ['UpdateBranch', 'Update'] })
            .toPromise());
    },
    async viewBranchAsync({ appId, name, limit, offset }) {
        return (0, client_1.withErrorHandlingAsync)(client_1.graphqlClient
            .query((0, graphql_tag_1.default) `
            query ViewBranchUpdates($appId: String!, $name: String!, $limit: Int!, $offset: Int!) {
              app {
                byId(appId: $appId) {
                  id
                  updateBranchByName(name: $name) {
                    id
                    name
                    updates(limit: $limit, offset: $offset) {
                      id
                      group
                      message
                      createdAt
                      runtimeVersion
                      platform
                      manifestFragment
                      actor {
                        id
                        ... on User {
                          username
                        }
                        ... on Robot {
                          firstName
                        }
                      }
                      branch {
                        id
                        name
                      }
                    }
                  }
                }
              }
            }
          `, {
            appId,
            name,
            limit,
            offset,
        }, { additionalTypenames: ['UpdateBranch', 'Update'] })
            .toPromise());
    },
};
