// eas-cli/build/commands/metadata/push.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const config_1 = require("@expo/config");
const core_1 = require("@oclif/core");
const configure_1 = require("../../build/configure");
const EasCommand_1 = tslib_1.__importDefault(require("../../commandUtils/EasCommand"));
const context_1 = require("../../credentials/context");
const log_1 = tslib_1.__importStar(require("../../log"));
const context_2 = require("../../metadata/context");
const errors_1 = require("../../metadata/errors");
const upload_1 = require("../../metadata/upload");
const projectUtils_1 = require("../../project/projectUtils");
const actions_1 = require("../../user/actions");
class MetadataPush extends EasCommand_1.default {
    async runAsync() {
        log_1.default.warn('EAS Metadata is in beta and subject to breaking changes.');
        const { flags } = await this.parse(MetadataPush);
        const projectDir = await (0, projectUtils_1.findProjectRootAsync)();
        const { exp } = (0, config_1.getConfig)(projectDir, { skipSDKVersionRequirement: true });
        await (0, projectUtils_1.getProjectIdAsync)(exp);
        await (0, configure_1.ensureProjectConfiguredAsync)({ projectDir, nonInteractive: false });
        const credentialsCtx = new context_1.CredentialsContext({
            exp,
            projectDir,
            user: await (0, actions_1.ensureLoggedInAsync)(),
            nonInteractive: false,
        });
        const metadataCtx = await (0, context_2.createMetadataContextAsync)({
            credentialsCtx,
            projectDir,
            exp,
            profileName: flags.profile,
        });
        try {
            const { appleLink } = await (0, upload_1.uploadMetadataAsync)(metadataCtx);
            log_1.default.addNewLineIfNone();
            log_1.default.log(`🎉 Store configuration is synced with the app stores.

${(0, log_1.learnMore)(appleLink, { learnMoreMessage: 'See the changes in App Store Connect' })}`);
        }
        catch (error) {
            (0, errors_1.handleMetadataError)(error);
        }
    }
}
exports.default = MetadataPush;
MetadataPush.description = 'sync the local store configuration to the app stores';
MetadataPush.flags = {
    profile: core_1.Flags.string({
        description: 'Name of the submit profile from eas.json. Defaults to "production" if defined in eas.json.',
    }),
};
