// react-native-beautiful-timeline/build/dist/components/Card/Card.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const React = (0, tslib_1.__importStar)(require("react"));
const prop_types_1 = (0, tslib_1.__importDefault)(require("prop-types"));
const react_native_1 = require("react-native");
const moment_1 = (0, tslib_1.__importDefault)(require("moment"));
const react_native_androw_1 = (0, tslib_1.__importDefault)(require("react-native-androw"));
const react_native_helpers_1 = require("@freakycoder/react-native-helpers");
/**
 * ? Local Imports
 */
const Card_style_1 = (0, tslib_1.__importStar)(require("./Card.style"));
const Card = (props) => {
    const { data, date, title, isCard, subtitle, dateStyle, titleStyle, shadowColor, subtitleStyle, dateFontColor, titleFontColor, subtitleFontColor, cardBackgroundColor, } = props;
    return (<react_native_androw_1.default style={[Card_style_1.default.container, react_native_helpers_1.isAndroid && (0, Card_style_1._shadowStyle)(isCard, shadowColor)]}>
      <react_native_androw_1.default style={(0, Card_style_1._cardContainer)(isCard, shadowColor, cardBackgroundColor)}>
        <react_native_1.View style={Card_style_1.default.cardContainerGlue}>
          <react_native_1.Text numberOfLines={1} style={titleStyle || (0, Card_style_1._titleStyle)(titleFontColor)}>
            {data.title}
          </react_native_1.Text>
          <react_native_1.Text numberOfLines={2} style={subtitleStyle || (0, Card_style_1._subtitleStyle)(subtitleFontColor)}>
            {data.subtitle}
          </react_native_1.Text>
        </react_native_1.View>
      </react_native_androw_1.default>
      <react_native_1.Text numberOfLines={1} style={dateStyle || (0, Card_style_1._dateStyle)(dateFontColor, isCard)}>
        {(0, moment_1.default)(data.date).format("DD ddd, HH:mm")}
      </react_native_1.Text>
    </react_native_androw_1.default>);
};
Card.propTypes = {
    date: prop_types_1.default.string,
    isCard: prop_types_1.default.bool,
    title: prop_types_1.default.string,
    subtitle: prop_types_1.default.string,
    shadowColor: prop_types_1.default.string,
    dateFontColor: prop_types_1.default.string,
    titleFontColor: prop_types_1.default.string,
    subtitleFontColor: prop_types_1.default.string,
    subtitleFontFamily: prop_types_1.default.string,
    cardBackgroundColor: prop_types_1.default.string,
};
Card.defaultProps = {
    isCard: true,
    shadowColor: "#000",
    date: "Tue 16, 19:09",
    dateFontColor: "#ccc",
    titleFontColor: "#556084",
    cardBackgroundColor: "#fff",
    subtitleFontColor: "#8c93ab",
    title: "React Native Beautiful Timeline",
    subtitle: "Etiam volutpat ligula metus, quis.",
};
exports.default = Card;
