// react-native-beautiful-timeline/build/dist/components/Card/Card.style.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports._cardContainer = exports._shadowStyle = exports._titleStyle = exports._subtitleStyle = exports._dateStyle = void 0;
const react_native_1 = require("react-native");
const _dateStyle = (color, isCard) => ({
    color: color,
    fontSize: 10,
    marginLeft: 32,
    marginTop: isCard ? 8 : 0,
});
exports._dateStyle = _dateStyle;
const _subtitleStyle = (color) => ({
    color: color,
    fontSize: 12,
    marginTop: 8,
    fontWeight: "600",
});
exports._subtitleStyle = _subtitleStyle;
const _titleStyle = (color) => ({
    color: color,
    fontSize: 14,
    fontWeight: "bold",
});
exports._titleStyle = _titleStyle;
const _shadowStyle = (isCard, shadowColor) => {
    return isCard
        ? {
            backgroundColor: "transparent",
            shadowColor,
            shadowRadius: 7,
            shadowOpacity: 0.09,
            shadowOffset: {
                width: 0,
                height: 3,
            },
        }
        : {
            backgroundColor: "transparent",
        };
};
exports._shadowStyle = _shadowStyle;
const _cardContainer = (isCard, shadowColor, backgroundColor) => {
    return [
        {
            marginTop: -5,
            paddingTop: 12,
            marginLeft: 24,
            borderRadius: 12,
            flexDirection: "row",
            paddingBottom: isCard ? 12 : 6,
        },
        isCard
            ? {
                shadowColor,
                backgroundColor,
                shadowRadius: 7,
                shadowOpacity: 0.05,
                shadowOffset: {
                    width: 0,
                    height: 3,
                },
            }
            : {},
    ];
};
exports._cardContainer = _cardContainer;
exports.default = react_native_1.StyleSheet.create({
    container: {
        width: "85%",
        paddingTop: 12,
        paddingLeft: 16,
        paddingBottom: 3,
        borderRadius: 322,
        alignSelf: "baseline",
        flexDirection: "column",
    },
    cardContainerGlue: {
        width: "100%",
        paddingLeft: 16,
    },
});
